cartridge/roles/ddl-manager.lua

```diff
@@ -0,0 +1,283 @@
+        local err = string.format(
+            'Bad argument #1 to set_clusterwide_schema_lua' ..
+            ' (?table expected, got %s)', type(schema_lua)
+        )
+        error(err, 2)
+    else
+        patch = {[_section_name] = yaml.encode(schema_lua)}
+    end
+
+    return cartridge.config_patch_clusterwide(patch)
+end
+
+--- Validate schema passed as a YAML string.
+--
+-- @function check_schema_yaml
+-- @tparam string schema_yml
+-- @treturn[1] boolean `true`
+-- @treturn[2] nil
+-- @treturn[2] table Error object
+local function check_schema_yaml(schema_yml)
+    if schema_yml == nil then
+        return true
+    elseif type(schema_yml) ~= 'string' then
+        return nil, CheckSchemaError:new(
+            'Bad argument #1 to check_schema_yaml' ..
+            ' (?string expected, got %s)', type(schema_yml)
+        )
+    end
+
+    local schema_lua, err = DecodeYamlError:pcall(yaml.decode, schema_yml)
+    if err ~= nil then
+        return nil, err
+    elseif schema_lua == nil then
+        return true
+    end
+
+    if not failover.is_leader() then
+        return cartridge.rpc_call(
+            'ddl-manager', 'check_schema_yaml', {schema_yml},
+            {prefer_local = true, leader_only = true}
+        )
+    end
+
+    local ok, err = ddl.check_schema(schema_lua)
+    if not ok then
+        return nil, CheckSchemaError:new(err)
+    end
+
+    return true
+end
+
+--- Validate schema passed as a Lua table.
+--
+-- @function check_schema_lua
+-- @tparam string schema_lua
+-- @treturn[1] boolean `true`
+-- @treturn[2] nil
+-- @treturn[2] table Error object
+local function check_schema_lua(schema_lua)
+    if schema_lua == nil then
+        return true
+    elseif type(schema_lua) ~= 'table' then
+        return nil, CheckSchemaError:new(
+            'Bad argument #1 to check_schema_lua' ..
+            ' (?table expected, got %s)', type(schema_lua)
+        )
+    end
+
+    -- Always perform encode + decode because
+    -- that's how `set_schema_lua` works
+    local ok, schema_yml = pcall(yaml.encode, schema_lua)
+    if not ok then
+        return nil, CheckSchemaError:new(
+            'Encoding YAML failed: %s', schema_yml
+        )
+    end
+
+    return cartridge.rpc_call(
+        'ddl-manager', 'check_schema_yaml', {schema_yml},
+        {prefer_local = true, leader_only = true}
+    )
+end
+
+return {
+    role_name = 'ddl-manager',
+    permanent = true,
+    _section_name = _section_name,
+    _example_schema = _example_schema,
+
+    init = init,
+    stop = stop,
+    validate_config = validate_config,
+    apply_config = apply_config,
+
+    get_clusterwide_schema_yaml = get_clusterwide_schema_yaml,
+    get_clusterwide_schema_lua = get_clusterwide_schema_lua,
+
+    set_clusterwide_schema_yaml = set_clusterwide_schema_yaml,
+    set_clusterwide_schema_lua = set_clusterwide_schema_lua,
+
+    check_schema_yaml = check_schema_yaml,
+    check_schema_lua = check_schema_lua,
```

Comment:
I think, it is not "lua" it is actually a json. So maybe it is worth adding "check_schema_json" and "check_schema" without suffix?

Reply:
It's not a json, but a Lua table. And if you want to check json-encoded string you can use `_yaml` functions (because yaml is a superset of json format)